Stack OverFlow Error

runtime: goroutine stack exceeds 1000000000-byte limit
fatal error: stack overflow

runtime stack:
runtime.throw(0x34b04a2, 0xe)
	/usr/local/go/src/runtime/panic.go:617 +0x72
runtime.newstack()
	/usr/local/go/src/runtime/stack.go:1041 +0x6f0
runtime.morestack()
	/usr/local/go/src/runtime/asm_amd64.s:429 +0x8f

goroutine 1 [running]:
runtime.rawstringtmp(0xc03c0004b8, 0x12, 0x0, 0x0, 0x0, 0x0, 0x0)
	/usr/local/go/src/runtime/string.go:118 +0xa0 fp=0xc03c000348 sp=0xc03c000340 pc=0x8e6280
runtime.concatstrings(0xc03c0004b8, 0xc03c000428, 0x3, 0x3, 0x0, 0x0)
	/usr/local/go/src/runtime/string.go:49 +0xae fp=0xc03c0003e0 sp=0xc03c000348 pc=0x8e5cde
runtime.concatstring3(0xc03c0004b8, 0x34acdc5, 0x7, 0x34ab895, 0x1, 0xc00092f8d4, 0xa, 0x0, 0x0)
	/usr/local/go/src/runtime/string.go:62 +0x47 fp=0xc03c000420 sp=0xc03c0003e0 pc=0x8e5fc7
github.com/tinygo-org/tinygo/compiler.(*Compiler).getLLVMType(0xc000178000, 0x352e3c0, 0xc002a3f6b0, 0x0)
	/home/circleci/project/compiler/compiler.go:500 +0x1021 fp=0xc03c000538 sp=0xc03c000420 pc=0xb7ed91
github.com/tinygo-org/tinygo/compiler.(*Compiler).getDIType(0xc000178000, 0x352e3c0, 0xc002a3f6b0, 0xa)
	/home/circleci/project/compiler/compiler.go:580 +0x5d fp=0xc03c000b98 sp=0xc03c000538 pc=0xb7f05d
github.com/tinygo-org/tinygo/compiler.(*Compiler).getDIType(0xc000178000, 0x352e340, 0xc0000e0af0, 0x604da60)
	/home/circleci/project/compiler/compiler.go:647 +0x1801 fp=0xc03c0011f8 sp=0xc03c000b98 pc=0xb80801
github.com/tinygo-org/tinygo/compiler.(*Compiler).getDIType(0xc000178000, 0x352e3c0, 0xc0000944b0, 0x4)
	/home/circleci/project/compiler/compiler.go:652 +0x2147 fp=0xc03c001858 sp=0xc03c0011f8 pc=0xb81147
github.com/tinygo-org/tinygo/compiler.(*Compiler).getDIType(0xc000178000, 0x352e4c0, 0xc003212db0, 0x60b19f0)
	/home/circleci/project/compiler/compiler.go:730 +0x349 fp=0xc03c001eb8 sp=0xc03c001858 pc=0xb7f349
github.com/tinygo-org/tinygo/compiler.(*Compiler).getDIType(0xc000178000, 0x352e3c0, 0xc003212d80, 0x60b1a10)
	/home/circleci/project/compiler/compiler.go:652 +0x2147 fp=0xc03c002518 sp=0xc03c001eb8 pc=0xb81147
github.com/tinygo-org/tinygo/compiler.(*Compiler).getDIType(0xc000178000, 0x352e400, 0xc000b71380, 0x60b1a30)
	/home/circleci/project/compiler/compiler.go:657 +0x5d0 fp=0xc03c002b78 sp=0xc03c002518 pc=0xb7f5d0
github.com/tinygo-org/tinygo/compiler.(*Compiler).getDIType(0xc000178000, 0x352e400, 0xc0052aa5c0, 0xc01cba1980)
	/home/circleci/project/compiler/compiler.go:657 +0x5d0 fp=0xc03c0031d8 sp=0xc03c002b78 pc=0xb7f5d0
github.com/tinygo-org/tinygo/compiler.(*Compiler).getDIType(0xc000178000, 0x352e480, 0xc000b712e0, 0xc)
	/home/circleci/project/compiler/compiler.go:697 +0xba1 fp=0xc03c003838 sp=0xc03c0031d8 pc=0xb7fba1
github.com/tinygo-org/tinygo/compiler.(*Compiler).getDIType(0xc000178000, 0x352e4c0, 0xc003212db0, 0x60b19f0)
	/home/circleci/project/compiler/compiler.go:730 +0x349 fp=0xc03c003e98 sp=0xc03c003838 pc=0xb7f349
github.com/tinygo-org/tinygo/compiler.(*Compiler).getDIType(0xc000178000, 0x352e3c0, 0xc003212d80, 0x60b1a10)
	/home/circleci/project/compiler/compiler.go:652 +0x2147 fp=0xc03c0044f8 sp=0xc03c003e98 pc=0xb81147
github.com/tinygo-org/tinygo/compiler.(*Compiler).getDIType(0xc000178000, 0x352e400, 0xc000b71380, 0x60b1a30)
	/home/circleci/project/compiler/compiler.go:657 +0x5d0 fp=0xc03c004b58 sp=0xc03c0044f8 pc=0xb7f5d0
github.com/tinygo-org/tinygo/compiler.(*Compiler).getDIType(0xc000178000, 0x352e400, 0xc0052aa400, 0xc01cba1800)
	/home/circleci/project/compiler/compiler.go:657 +0x5d0 fp=0xc03c0051b8 sp=0xc03c004b58 pc=0xb7f5d0
github.com/tinygo-org/tinygo/compiler.(*Compiler).getDIType(0xc000178000, 0x352e480, 0xc000b712e0, 0xc)
	/home/circleci/project/compiler/compiler.go:697 +0xba1 fp=0xc03c005818 sp=0xc03c0051b8 pc=0xb7fba1
github.com/tinygo-org/tinygo/compiler.(*Compiler).getDIType(0xc000178000, 0x352e4c0, 0xc003212db0, 0x60b19f0)
	/home/circleci/project/compiler/compiler.go:730 +0x349 fp=0xc03c005e78 sp=0xc03c005818 pc=0xb7f349
github.com/tinygo-org/tinygo/compiler.(*Compiler).getDIType(0xc000178000, 0x352e3c0, 0xc003212d80, 0x60b1a10)
	/home/circleci/project/compiler/compiler.go:652 +0x2147 fp=0xc03c0064d8 sp=0xc03c005e78 pc=0xb81147
github.com/tinygo-org/tinygo/compiler.(*Compiler).getDIType(0xc000178000, 0x352e400, 0xc000b71380, 0x60b1a30)
	/home/circleci/project/compiler/compiler.go:657 +0x5d0 fp=0xc03c006b38 sp=0xc03c0064d8 pc=0xb7f5d0
github.com/tinygo-org/tinygo/compiler.(*Compiler).getDIType(0xc000178000, 0x352e400, 0xc0052aa240, 0xc01cba1680)
	/home/circleci/project/compiler/compiler.go:657 +0x5d0 fp=0xc03c007198 sp=0xc03c006b38 pc=0xb7f5d0
github.com/tinygo-org/tinygo/compiler.(*Compiler).getDIType(0xc000178000, 0x352e480, 0xc000b712e0, 0xc)
	/home/circleci/project/compiler/compiler.go:697 +0xba1 fp=0xc03c0077f8 sp=0xc03c007198 pc=0xb7fba1
github.com/tinygo-org/tinygo/compiler.(*Compiler).getDIType(0xc000178000, 0x352e4c0, 0xc003212db0, 0x60b19f0)
	/home/circleci/project/compiler/compiler.go:730 +0x349 fp=0xc03c007e58 sp=0xc03c0077f8 pc=0xb7f349
github.com/tinygo-org/tinygo/compiler.(*Compiler).getDIType(0xc000178000, 0x352e3c0, 0xc003212d80, 0x60b1a10)
	/home/circleci/project/compiler/compiler.go:652 +0x2147 fp=0xc03c0084b8 sp=0xc03c007e58 pc=0xb81147
github.com/tinygo-org/tinygo/compiler.(*Compiler).getDIType(0xc000178000, 0x352e400, 0xc000b71380, 0x60b1a30)
	/home/circleci/project/compiler/compiler.go:657 +0x5d0 fp=0xc03c008b18 sp=0xc03c0084b8 pc=0xb7f5d0
github.com/tinygo-org/tinygo/compiler.(*Compiler).getDIType(0xc000178000, 0x352e400, 0xc0052aa080, 0xc01cba14c0)
	/home/circleci/project/compiler/compiler.go:657 +0x5d0 fp=0xc03c009178 sp=0xc03c008b18 pc=0xb7f5d0
github.com/tinygo-org/tinygo/compiler.(*Compiler).getDIType(0xc000178000, 0x352e480, 0xc000b712e0, 0xc)
	/home/circleci/project/compiler/compiler.go:697 +0xba1 fp=0xc03c0097d8 sp=0xc03c009178 pc=0xb7fba1
github.com/tinygo-org/tinygo/compiler.(*Compiler).getDIType(0xc000178000, 0x352e4c0, 0xc003212db0, 0x60b19f0)
	/home/circleci/project/compiler/compiler.go:730 +0x349 fp=0xc03c009e38 sp=0xc03c0097d8 pc=0xb7f349
github.com/tinygo-org/tinygo/compiler.(*Compiler).getDIType(0xc000178000, 0x352e3c0, 0xc003212d80, 0x60b1a10)
	/home/circleci/project/compiler/compiler.go:652 +0x2147 fp=0xc03c00a498 sp=0xc03c009e38 pc=0xb81147
github.com/tinygo-org/tinygo/compiler.(*Compiler).getDIType(0xc000178000, 0x352e400, 0xc000b71380, 0x60b1a30)
	/home/circleci/project/compiler/compiler.go:657 +0x5d0 fp=0xc03c00aaf8 sp=0xc03c00a498 pc=0xb7f5d0
github.com/tinygo-org/tinygo/compiler.(*Compiler).getDIType(0xc000178000, 0x352e400, 0xc0052a9ec0, 0xc01cba1340)
	/home/circleci/project/compiler/compiler.go:657 +0x5d0 fp=0xc03c00b158 sp=0xc03c00aaf8 pc=0xb7f5d0
github.com/tinygo-org/tinygo/compiler.(*Compiler).getDIType(0xc000178000, 0x352e480, 0xc000b712e0, 0xc)
	/home/circleci/project/compiler/compiler.go:697 +0xba1 fp=0xc03c00b7b8 sp=0xc03c00b158 pc=0xb7fba1
github.com/tinygo-org/tinygo/compiler.(*Compiler).getDIType(0xc000178000, 0x352e4c0, 0xc003212db0, 0x60b19f0)
	/home/circleci/project/compiler/compiler.go:730 +0x349 fp=0xc03c00be18 sp=0xc03c00b7b8 pc=0xb7f349
github.com/tinygo-org/tinygo/compiler.(*Compiler).getDIType(0xc000178000, 0x352e3c0, 0xc003212d80, 0x60b1a10)
	/home/circleci/project/compiler/compiler.go:652 +0x2147 fp=0xc03c00c478 sp=0xc03c00be18 pc=0xb81147
github.com/tinygo-org/tinygo/compiler.(*Compiler).getDIType(0xc000178000, 0x352e400, 0xc000b71380, 0x60b1a30)
	/home/circleci/project/compiler/compiler.go:657 +0x5d0 fp=0xc03c00cad8 sp=0xc03c00c478 pc=0xb7f5d0
github.com/tinygo-org/tinygo/compiler.(*Compiler).getDIType(0xc000178000, 0x352e400, 0xc0052a9d00, 0xc01cba11c0)
	/home/circleci/project/compiler/compiler.go:657 +0x5d0 fp=0xc03c00d138 sp=0xc03c00cad8 pc=0xb7f5d0
github.com/tinygo-org/tinygo/compiler.(*Compiler).getDIType(0xc000178000, 0x352e480, 0xc000b712e0, 0xc)
	/home/circleci/project/compiler/compiler.go:697 +0xba1 fp=0xc03c00d798 sp=0xc03c00d138 pc=0xb7fba1
github.com/tinygo-org/tinygo/compiler.(*Compiler).getDIType(0xc000178000, 0x352e4c0, 0xc003212db0, 0x60b19f0)
	/home/circleci/project/compiler/compiler.go:730 +0x349 fp=0xc03c00ddf8 sp=0xc03c00d798 pc=0xb7f349
github.com/tinygo-org/tinygo/compiler.(*Compiler).getDIType(0xc000178000, 0x352e3c0, 0xc003212d80, 0x60b1a10)
	/home/circleci/project/compiler/compiler.go:652 +0x2147 fp=0xc03c00e458 sp=0xc03c00ddf8 pc=0xb81147
github.com/tinygo-org/tinygo/compiler.(*Compiler).getDIType(0xc000178000, 0x352e400, 0xc000b71380, 0x60b1a30)
	/home/circleci/project/compiler/compiler.go:657 +0x5d0 fp=0xc03c00eab8 sp=0xc03c00e458 pc=0xb7f5d0
github.com/tinygo-org/tinygo/compiler.(*Compiler).getDIType(0xc000178000, 0x352e400, 0xc0052a9b40, 0xc01cba1040)
	/home/circleci/project/compiler/compiler.go:657 +0x5d0 fp=0xc03c00f118 sp=0xc03c00eab8 pc=0xb7f5d0
github.com/tinygo-org/tinygo/compiler.(*Compiler).getDIType(0xc000178000, 0x352e480, 0xc000b712e0, 0xc)
	/home/circleci/project/compiler/compiler.go:697 +0xba1 fp=0xc03c00f778 sp=0xc03c00f118 pc=0xb7fba1
github.com/tinygo-org/tinygo/compiler.(*Compiler).getDIType(0xc000178000, 0x352e4c0, 0xc003212db0, 0x60b19f0)
	/home/circleci/project/compiler/compiler.go:730 +0x349 fp=0xc03c00fdd8 sp=0xc03c00f778 pc=0xb7f349
github.com/tinygo-org/tinygo/compiler.(*Compiler).getDIType(0xc000178000, 0x352e3c0, 0xc003212d80, 0x60b1a10)
	/home/circleci/project/compiler/compiler.go:652 +0x2147 fp=0xc03c010438 sp=0xc03c00fdd8 pc=0xb81147
github.com/tinygo-org/tinygo/compiler.(*Compiler).getDIType(0xc000178000, 0x352e400, 0xc000b71380, 0x60b1a30)
	/home/circleci/project/compiler/compiler.go:657 +0x5d0 fp=0xc03c010a98 sp=0xc03c010438 pc=0xb7f5d0
github.com/tinygo-org/tinygo/compiler.(*Compiler).getDIType(0xc000178000, 0x352e400, 0xc0052a9980, 0xc01cba0e80)
	/home/circleci/project/compiler/compiler.go:657 +0x5d0 fp=0xc03c0110f8 sp=0xc03c010a98 pc=0xb7f5d0
github.com/tinygo-org/tinygo/compiler.(*Compiler).getDIType(0xc000178000, 0x352e480, 0xc000b712e0, 0xc)
	/home/circleci/project/compiler/compiler.go:697 +0xba1 fp=0xc03c011758 sp=0xc03c0110f8 pc=0xb7fba1
github.com/tinygo-org/tinygo/compiler.(*Compiler).getDIType(0xc000178000, 0x352e4c0, 0xc003212db0, 0x60b19f0)
	/home/circleci/project/compiler/compiler.go:730 +0x349 fp=0xc03c011db8 sp=0xc03c011758 pc=0xb7f349
github.com/tinygo-org/tinygo/compiler.(*Compiler).getDIType(0xc000178000, 0x352e3c0, 0xc003212d80, 0x60b1a10)
	/home/circleci/project/compiler/compiler.go:652 +0x2147 fp=0xc03c012418 sp=0xc03c011db8 pc=0xb81147
github.com/tinygo-org/tinygo/compiler.(*Compiler).getDIType(0xc000178000, 0x352e400, 0xc000b71380, 0x60b1a30)
	/home/circleci/project/compiler/compiler.go:657 +0x5d0 fp=0xc03c012a78 sp=0xc03c012418 pc=0xb7f5d0
github.com/tinygo-org/tinygo/compiler.(*Compiler).getDIType(0xc000178000, 0x352e400, 0xc0052a97c0, 0xc01cba0d00)
	/home/circleci/project/compiler/compiler.go:657 +0x5d0 fp=0xc03c0130d8 sp=0xc03c012a78 pc=0xb7f5d0
github.com/tinygo-org/tinygo/compiler.(*Compiler).getDIType(0xc000178000, 0x352e480, 0xc000b712e0, 0xc)
	/home/circleci/project/compiler/compiler.go:697 +0xba1 fp=0xc03c013738 sp=0xc03c0130d8 pc=0xb7fba1
github.com/tinygo-org/tinygo/compiler.(*Compiler).getDIType(0xc000178000, 0x352e4c0, 0xc003212db0, 0x60b19f0)
	/home/circleci/project/compiler/compiler.go:730 +0x349 fp=0xc03c013d98 sp=0xc03c013738 pc=0xb7f349
github.com/tinygo-org/tinygo/compiler.(*Compiler).getDIType(0xc000178000, 0x352e3c0, 0xc003212d80, 0x60b1a10)
	/home/circleci/project/compiler/compiler.go:652 +0x2147 fp=0xc03c0143f8 sp=0xc03c013d98 pc=0xb81147
github.com/tinygo-org/tinygo/compiler.(*Compiler).getDIType(0xc000178000, 0x352e400, 0xc000b71380, 0x60b1a30)
	/home/circleci/project/compiler/compiler.go:657 +0x5d0 fp=0xc03c014a58 sp=0xc03c0143f8 pc=0xb7f5d0
github.com/tinygo-org/tinygo/compiler.(*Compiler).getDIType(0xc000178000, 0x352e400, 0xc0052a9600, 0xc01cba0b80)
	/home/circleci/project/compiler/compiler.go:657 +0x5d0 fp=0xc03c0150b8 sp=0xc03c014a58 pc=0xb7f5d0
github.com/tinygo-org/tinygo/compiler.(*Compiler).getDIType(0xc000178000, 0x352e480, 0xc000b712e0, 0xc)
	/home/circleci/project/compiler/compiler.go:697 +0xba1 fp=0xc03c015718 sp=0xc03c0150b8 pc=0xb7fba1
github.com/tinygo-org/tinygo/compiler.(*Compiler).getDIType(0xc000178000, 0x352e4c0, 0xc003212db0, 0x60b19f0)
	/home/circleci/project/compiler/compiler.go:730 +0x349 fp=0xc03c015d78 sp=0xc03c015718 pc=0xb7f349
github.com/tinygo-org/tinygo/compiler.(*Compiler).getDIType(0xc000178000, 0x352e3c0, 0xc003212d80, 0x60b1a10)
	/home/circleci/project/compiler/compiler.go:652 +0x2147 fp=0xc03c0163d8 sp=0xc03c015d78 pc=0xb81147
github.com/tinygo-org/tinygo/compiler.(*Compiler).getDIType(0xc000178000, 0x352e400, 0xc000b71380, 0x60b1a30)
	/home/circleci/project/compiler/compiler.go:657 +0x5d0 fp=0xc03c016a38 sp=0xc03c0163d8 pc=0xb7f5d0
github.com/tinygo-org/tinygo/compiler.(*Compiler).getDIType(0xc000178000, 0x352e400, 0xc0052a9440, 0xc01cba0a00)
	/home/circleci/project/compiler/compiler.go:657 +0x5d0 fp=0xc03c017098 sp=0xc03c016a38 pc=0xb7f5d0
github.com/tinygo-org/tinygo/compiler.(*Compiler).getDIType(0xc000178000, 0x352e480, 0xc000b712e0, 0xc)
	/home/circleci/project/compiler/compiler.go:697 +0xba1 fp=0xc03c0176f8 sp=0xc03c017098 pc=0xb7fba1
github.com/tinygo-org/tinygo/compiler.(*Compiler).getDIType(0xc000178000, 0x352e4c0, 0xc003212db0, 0x60b19f0)
	/home/circleci/project/compiler/compiler.go:730 +0x349 fp=0xc03c017d58 sp=0xc03c0176f8 pc=0xb7f349
github.com/tinygo-org/tinygo/compiler.(*Compiler).getDIType(0xc000178000, 0x352e3c0, 0xc003212d80, 0x60b1a10)
	/home/circleci/project/compiler/compiler.go:652 +0x2147 fp=0xc03c0183b8 sp=0xc03c017d58 pc=0xb81147
github.com/tinygo-org/tinygo/compiler.(*Compiler).getDIType(0xc000178000, 0x352e400, 0xc000b71380, 0x60b1a30)
	/home/circleci/project/compiler/compiler.go:657 +0x5d0 fp=0xc03c018a18 sp=0xc03c0183b8 pc=0xb7f5d0
github.com/tinygo-org/tinygo/compiler.(*Compiler).getDIType(0xc000178000, 0x352e400, 0xc0052a9280, 0xc01cba0840)
	/home/circleci/project/compiler/compiler.go:657 +0x5d0 fp=0xc03c019078 sp=0xc03c018a18 pc=0xb7f5d0
github.com/tinygo-org/tinygo/compiler.(*Compiler).getDIType(0xc000178000, 0x352e480, 0xc000b712e0, 0xc)
	/home/circleci/project/compiler/compiler.go:697 +0xba1 fp=0xc03c0196d8 sp=0xc03c019078 pc=0xb7fba1
github.com/tinygo-org/tinygo/compiler.(*Compiler).getDIType(0xc000178000, 0x352e4c0, 0xc003212db0, 0x60b19f0)
	/home/circleci/project/compiler/compiler.go:730 +0x349 fp=0xc03c019d38 sp=0xc03c0196d8 pc=0xb7f349
github.com/tinygo-org/tinygo/compiler.(*Compiler).getDIType(0xc000178000, 0x352e3c0, 0xc003212d80, 0x60b1a10)
	/home/circleci/project/compiler/compiler.go:652 +0x2147 fp=0xc03c01a398 sp=0xc03c019d38 pc=0xb81147
github.com/tinygo-org/tinygo/compiler.(*Compiler).getDIType(0xc000178000, 0x352e400, 0xc000b71380, 0x60b1a30)
	/home/circleci/project/compiler/compiler.go:657 +0x5d0 fp=0xc03c01a9f8 sp=0xc03c01a398 pc=0xb7f5d0
github.com/tinygo-org/tinygo/compiler.(*Compiler).getDIType(0xc000178000, 0x352e400, 0xc0052a90c0, 0xc01cba06c0)
	/home/circleci/project/compiler/compiler.go:657 +0x5d0 fp=0xc03c01b058 sp=0xc03c01a9f8 pc=0xb7f5d0
github.com/tinygo-org/tinygo/compiler.(*Compiler).getDIType(0xc000178000, 0x352e480, 0xc000b712e0, 0xc)
	/home/circleci/project/compiler/compiler.go:697 +0xba1 fp=0xc03c01b6b8 sp=0xc03c01b058 pc=0xb7fba1
github.com/tinygo-org/tinygo/compiler.(*Compiler).getDIType(0xc000178000, 0x352e4c0, 0xc003212db0, 0x60b19f0)
	/home/circleci/project/compiler/compiler.go:730 +0x349 fp=0xc03c01bd18 sp=0xc03c01b6b8 pc=0xb7f349
github.com/tinygo-org/tinygo/compiler.(*Compiler).getDIType(0xc000178000, 0x352e3c0, 0xc003212d80, 0x60b1a10)
	/home/circleci/project/compiler/compiler.go:652 +0x2147 fp=0xc03c01c378 sp=0xc03c01bd18 pc=0xb81147
github.com/tinygo-org/tinygo/compiler.(*Compiler).getDIType(0xc000178000, 0x352e400, 0xc000b71380, 0x60b1a30)
	/home/circleci/project/compiler/compiler.go:657 +0x5d0 fp=0xc03c01c9d8 sp=0xc03c01c378 pc=0xb7f5d0
github.com/tinygo-org/tinygo/compiler.(*Compiler).getDIType(0xc000178000, 0x352e400, 0xc0052a8f00, 0xc01cba0540)
	/home/circleci/project/compiler/compiler.go:657 +0x5d0 fp=0xc03c01d038 sp=0xc03c01c9d8 pc=0xb7f5d0
github.com/tinygo-org/tinygo/compiler.(*Compiler).getDIType(0xc000178000, 0x352e480, 0xc000b712e0, 0xc)
	/home/circleci/project/compiler/compiler.go:697 +0xba1 fp=0xc03c01d698 sp=0xc03c01d038 pc=0xb7fba1
github.com/tinygo-org/tinygo/compiler.(*Compiler).getDIType(0xc000178000, 0x352e4c0, 0xc003212db0, 0x60b19f0)
	/home/circleci/project/compiler/compiler.go:730 +0x349 fp=0xc03c01dcf8 sp=0xc03c01d698 pc=0xb7f349
github.com/tinygo-org/tinygo/compiler.(*Compiler).getDIType(0xc000178000, 0x352e3c0, 0xc003212d80, 0x60b1a10)
	/home/circleci/project/compiler/compiler.go:652 +0x2147 fp=0xc03c01e358 sp=0xc03c01dcf8 pc=0xb81147
github.com/tinygo-org/tinygo/compiler.(*Compiler).getDIType(0xc000178000, 0x352e400, 0xc000b71380, 0x60b1a30)
	/home/circleci/project/compiler/compiler.go:657 +0x5d0 fp=0xc03c01e9b8 sp=0xc03c01e358 pc=0xb7f5d0
github.com/tinygo-org/tinygo/compiler.(*Compiler).getDIType(0xc000178000, 0x352e400, 0xc0052a8d40, 0xc01cba03c0)
	/home/circleci/project/compiler/compiler.go:657 +0x5d0 fp=0xc03c01f018 sp=0xc03c01e9b8 pc=0xb7f5d0
github.com/tinygo-org/tinygo/compiler.(*Compiler).getDIType(0xc000178000, 0x352e480, 0xc000b712e0, 0xc)
	/home/circleci/project/compiler/compiler.go:697 +0xba1 fp=0xc03c01f678 sp=0xc03c01f018 pc=0xb7fba1
github.com/tinygo-org/tinygo/compiler.(*Compiler).getDIType(0xc000178000, 0x352e4c0, 0xc003212db0, 0x60b19f0)
	/home/circleci/project/compiler/compiler.go:730 +0x349 fp=0xc03c01fcd8 sp=0xc03c01f678 pc=0xb7f349
github.com/tinygo-org/tinygo/compiler.(*Compiler).getDIType(0xc000178000, 0x352e3c0, 0xc003212d80, 0x60b1a10)
	/home/circleci/project/compiler/compiler.go:652 +0x2147 fp=0xc03c020338 sp=0xc03c01fcd8 pc=0xb81147
github.com/tinygo-org/tinygo/compiler.(*Compiler).getDIType(0xc000178000, 0x352e400, 0xc000b71380, 0x60b1a30)
	/home/circleci/project/compiler/compiler.go:657 +0x5d0 fp=0xc03c020998 sp=0xc03c020338 pc=0xb7f5d0
github.com/tinygo-org/tinygo/compiler.(*Compiler).getDIType(0xc000178000, 0x352e400, 0xc0052a8b80, 0xc01cba0200)
	/home/circleci/project/compiler/compiler.go:657 +0x5d0 fp=0xc03c020ff8 sp=0xc03c020998 pc=0xb7f5d0
github.com/tinygo-org/tinygo/compiler.(*Compiler).getDIType(0xc000178000, 0x352e480, 0xc000b712e0, 0xc)
	/home/circleci/project/compiler/compiler.go:697 +0xba1 fp=0xc03c021658 sp=0xc03c020ff8 pc=0xb7fba1
github.com/tinygo-org/tinygo/compiler.(*Compiler).getDIType(0xc000178000, 0x352e4c0, 0xc003212db0, 0x60b19f0)
	/home/circleci/project/compiler/compiler.go:730 +0x349 fp=0xc03c021cb8 sp=0xc03c021658 pc=0xb7f349
github.com/tinygo-org/tinygo/compiler.(*Compiler).getDIType(0xc000178000, 0x352e3c0, 0xc003212d80, 0x60b1a10)
	/home/circleci/project/compiler/compiler.go:652 +0x2147 fp=0xc03c022318 sp=0xc03c021cb8 pc=0xb81147
github.com/tinygo-org/tinygo/compiler.(*Compiler).getDIType(0xc000178000, 0x352e400, 0xc000b71380, 0x60b1a30)
	/home/circleci/project/compiler/compiler.go:657 +0x5d0 fp=0xc03c022978 sp=0xc03c022318 pc=0xb7f5d0
github.com/tinygo-org/tinygo/compiler.(*Compiler).getDIType(0xc000178000, 0x352e400, 0xc0052a89c0, 0xc01cba0080)
	/home/circleci/project/compiler/compiler.go:657 +0x5d0 fp=0xc03c022fd8 sp=0xc03c022978 pc=0xb7f5d0
github.com/tinygo-org/tinygo/compiler.(*Compiler).getDIType(0xc000178000, 0x352e480, 0xc000b712e0, 0xc)
	/home/circleci/project/compiler/compiler.go:697 +0xba1 fp=0xc03c023638 sp=0xc03c022fd8 pc=0xb7fba1
github.com/tinygo-org/tinygo/compiler.(*Compiler).getDIType(0xc000178000, 0x352e4c0, 0xc003212db0, 0x60b19f0)
	/home/circleci/project/compiler/compiler.go:730 +0x349 fp=0xc03c023c98 sp=0xc03c023638 pc=0xb7f349
github.com/tinygo-org/tinygo/compiler.(*Compiler).getDIType(0xc000178000, 0x352e3c0, 0xc003212d80, 0x60b1a10)
	/home/circleci/project/compiler/compiler.go:652 +0x2147 fp=0xc03c0242f8 sp=0xc03c023c98 pc=0xb81147
github.com/tinygo-org/tinygo/compiler.(*Compiler).getDIType(0xc000178000, 0x352e400, 0xc000b71380, 0x60b1a30)
	/home/circleci/project/compiler/compiler.go:657 +0x5d0 fp=0xc03c024958 sp=0xc03c0242f8 pc=0xb7f5d0
github.com/tinygo-org/tinygo/compiler.(*Compiler).getDIType(0xc000178000, 0x352e400, 0xc0052a8800, 0xc01cb91f00)
	/home/circleci/project/compiler/compiler.go:657 +0x5d0 fp=0xc03c024fb8 sp=0xc03c024958 pc=0xb7f5d0
github.com/tinygo-org/tinygo/compiler.(*Compiler).getDIType(0xc000178000, 0x352e480, 0xc000b712e0, 0xc)
	/home/circleci/project/compiler/compiler.go:697 +0xba1 fp=0xc03c025618 sp=0xc03c024fb8 pc=0xb7fba1
github.com/tinygo-org/tinygo/compiler.(*Compiler).getDIType(0xc000178000, 0x352e4c0, 0xc003212db0, 0x60b19f0)
	/home/circleci/project/compiler/compiler.go:730 +0x349 fp=0xc03c025c78 sp=0xc03c025618 pc=0xb7f349
github.com/tinygo-org/tinygo/compiler.(*Compiler).getDIType(0xc000178000, 0x352e3c0, 0xc003212d80, 0x60b1a10)
	/home/circleci/project/compiler/compiler.go:652 +0x2147 fp=0xc03c0262d8 sp=0xc03c025c78 pc=0xb81147
github.com/tinygo-org/tinygo/compiler.(*Compiler).getDIType(0xc000178000, 0x352e400, 0xc000b71380, 0x60b1a30)
	/home/circleci/project/compiler/compiler.go:657 +0x5d0 fp=0xc03c026938 sp=0xc03c0262d8 pc=0xb7f5d0
...additional frames elided...

goroutine 19 [syscall, 1 minutes]:
os/signal.signal_recv(0x0)
	/usr/local/go/src/runtime/sigqueue.go:139 +0x9c
os/signal.loop()
	/usr/local/go/src/os/signal/signal_unix.go:23 +0x22
created by os/signal.init.0
	/usr/local/go/src/os/signal/signal_unix.go:29 +0x41

test code:
package main

import (
	"fmt"
)

type itemType string

const (
	BLOCK itemType = "block"
)

func test() map[itemType]string {
	testMap := map[itemType]string{}
	testMap[BLOCK] = "test"
	return testMap
}

func main() {
	fmt.Println(test()[BLOCK])
}

output:
bug.go:21:23: only strings, bools, ints, pointers or structs of bools/ints are supported as map keys, but got: bug.go.itemType

--- a/compiler/compiler.go
+++ b/compiler/compiler.go
@@ -1,6 +1,7 @@
 package compiler
 
 import (
+	"debug/dwarf"
 	"errors"
 	"fmt"
 	"go/ast"
@@ -91,6 +92,7 @@ type Compiler struct {
 	dibuilder               *llvm.DIBuilder
 	cu                      llvm.Metadata
 	difiles                 map[string]llvm.Metadata
+	ditypes                 map[types.Type]llvm.Metadata
 	machine                 llvm.TargetMachine
 	targetData              llvm.TargetData
 	intType                 llvm.Type
@@ -136,6 +138,7 @@ func NewCompiler(pkgName string, config Config) (*Compiler, error) {
 	c := &Compiler{
 		Config:  config,
 		difiles: make(map[string]llvm.Metadata),
+		ditypes: make(map[types.Type]llvm.Metadata),
 	}
 
 	target, err := llvm.GetTargetFromTriple(config.Triple)
@@ -596,6 +599,17 @@ func isPointer(typ types.Type) bool {
 
 // Get the DWARF type for this Go type.
 func (c *Compiler) getDIType(typ types.Type) llvm.Metadata {
+	if md, ok := c.ditypes[typ]; ok {
+		return md
+	}
+	md := c.createDIType(typ)
+	c.ditypes[typ] = md
+	return md
+}
+
+// createDIType creates a new DWARF type. Don't call this function directly,
+// call getDIType instead.
+func (c *Compiler) createDIType(typ types.Type) llvm.Metadata {
 	llvmType := c.getLLVMType(typ)
 	sizeInBytes := c.targetData.TypeAllocSize(llvmType)
 	switch typ := typ.(type) {
@@ -732,14 +746,17 @@ func (c *Compiler) getDIType(typ types.Type) llvm.Metadata {
 			},
 		})
 	case *types.Struct:
+		// Placeholder metadata node, to be replaced afterwards.
+		temporaryMDNode := c.dibuilder.CreateReplaceableCompositeType(llvm.Metadata{}, llvm.DIReplaceableCompositeType{
+			Tag:         dwarf.TagStructType,
+			SizeInBits:  sizeInBytes * 8,
+			AlignInBits: uint32(c.targetData.ABITypeAlignment(llvmType)) * 8,
+		})
+		c.ditypes[typ] = temporaryMDNode
 		elements := make([]llvm.Metadata, typ.NumFields())
 		for i := range elements {
 			field := typ.Field(i)
 			fieldType := field.Type()
-			if _, ok := fieldType.Underlying().(*types.Pointer); ok {
-				// XXX hack to avoid recursive types
-				fieldType = types.Typ[types.UnsafePointer]
-			}
 			llvmField := c.getLLVMType(fieldType)
 			elements[i] = c.dibuilder.CreateMemberType(llvm.Metadata{}, llvm.DIMemberType{
 				Name:         field.Name(),
@@ -749,11 +766,13 @@ func (c *Compiler) getDIType(typ types.Type) llvm.Metadata {
 				Type:         c.getDIType(fieldType),
 			})
 		}
-		return c.dibuilder.CreateStructType(llvm.Metadata{}, llvm.DIStructType{
+		md := c.dibuilder.CreateStructType(llvm.Metadata{}, llvm.DIStructType{
 			SizeInBits:  sizeInBytes * 8,
 			AlignInBits: uint32(c.targetData.ABITypeAlignment(llvmType)) * 8,
 			Elements:    elements,
 		})
+		temporaryMDNode.ReplaceAllUsesWith(md)
+		return md
 	default:
 		panic("unknown type while generating DWARF debug type: " + typ.String())
 	}

--- a/testdata/structs.go
+++ b/testdata/structs.go
@@ -60,6 +60,13 @@ type s8 struct {
 	b byte   // 1 element
 }
 
+// linked list (recursive type)
+type s9 struct {
+	n    int
+	next *s9
+	s    []*s9
+}
+
 func test0(s s0) {
 	println("test0")
 }
@@ -106,6 +113,10 @@ func test8(s s8) {
 	println("test8", len(s.a), cap(s.a), s.a[0], s.a[1], s.b)
 }
 
+func test9(s s9) {
+	println("test9", s.next.next)
+}
+
 func main() {
 	test0(s0{})
 	test1(s1{1})
@@ -119,4 +130,5 @@ func main() {
 	test6(s6{"foo", 5})
 	test7(s7{a: nil, b: 8})
 	test8(s8{[]byte{12, 13, 14}[:2], 6})
+	test9(s9{next: &s9{}})
 }

--- a/testdata/structs.txt
+++ b/testdata/structs.txt
@@ -9,3 +9,4 @@ test5 1 2 3
 test6 foo 3 5
 test7 (0:nil) 8
 test8 2 3 12 13 6
+test9 nil

